Add Ziccif and Ziccid extensions

ved-rivos opened the pull request:
If the Ziccif extension is implemented, main memory regions with both the cacheability and coherence PMAs must support instruction fetch, and any instruction fetches of naturally aligned power-of-2 sizes of at most min(ILEN,XLEN) bits are atomic. This fetch state machine is updated to be compliant to this extension. 

The Ziccid extension mandates more stringent requirements for consistency between instruction fetches and other memory accesses than those imposed by the base ISA. The primary intent of the Ziccid extension is to accelerate JIT compilation in multiprocessor systems.In straightforward implementations, maintaining coherence between instruction caches and the data-memory system suffices to satisfy this extension’s strictures. The Ziccid extension can be viewed as a means to codify the concept of instruction-cache coherence. This is primarily added as an ISA string to the model.

Update has been tested with the units tests.

Specifcation: https://github.com/aswaterman/riscv-misc/blob/main/isa/ziccid.adoc

aswaterman reviewed (approved):
LGTM, but I'm not one of the SAIL model experts.

ved-rivos commented:
@pmundkur - could you please help me with cause for the lean backend failure

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -12,12 +12,37 @@
-   * but any exceptions use the untransformed PC.
-   */
-  // TODO: Add a parameter to try_step() to allow forcing the fetch result and use that instead.
+function read2_at(vaddr : xlenbits) -> result(bitvector(16), FetchResult) = {
```
updated

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -12,12 +12,37 @@
+  }
+}
+
+function read4_at(vaddr : xlenbits) -> result(bitvector(32), FetchResult) = {
```
updated

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -12,12 +12,37 @@
+          }
+      }
+  }
+}
```
merged

ved-rivos commented on README.md:
```diff
@@ -125,6 +125,8 @@ For booting operating system images, see the information under the
 - Svbare extension for Bare mode virtual-memory translation
 - Physical Memory Protection (PMP)
+- Ziccif Extension for Instruction-Fetch Atomicity
+- Ziccid Extension for Instruction/Data Coherence and Consistency
```
done

ved-rivos commented on model/riscv_extensions.sail:
```diff
@@ -112,6 +112,14 @@ function clause hartSupports(Ext_Zihpm) = config extensions.Zihpm.supported
+// Instruction-Fetch Atomicity
+enum clause extension = Ext_Ziccif
+mapping clause extensionName = Ext_Ziccif <-> "ziccif"
+function clause hartSupports(Ext_Ziccif) = config extensions.Ziccif.supported
```
done

ved-rivos commented:
@jordancarlin Could you suggest how to resolve this failure?
```
/home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/ir/LeanRV64D/RiscvExtensions.setup.json --json
error: LeanRV64D/RiscvExtensions.lean:544:37: Unknown identifier `sys_enable_experimental_extensions`
error: Lean exited with code 1
error: build failed
```

jrtc27 commented on model/riscv_fetch.sail:
```diff
@@ -29,33 +40,44 @@ function fetch() -> FetchResult = {
+  } else {
+    // PC % 4 == 2:
+    // Step 1: atomic 2B at PC.
+    // If not compressed, Step 3: atomic 4B at PC+2; use only its low halfword.
```
Why does this need to be an atomic 4B fetch that throws away 2B? Isn't that indistinguishable, aside from if you're snooping the bus or fetching from side-effecting memory, from fetching 2B?

jrtc27 replied:
To me it seems like the sensible approach would be "if Ziccif is enabled and PC is aligned and ext_fetch_check_pc grants access to the entire 4 bytes then fetch 4 bytes atomically, otherwise use the old pattern (fetch 2 bytes and conditionally fetch a further 2 bytes)".

ved-rivos replied:
Please see the Ziccif specification - step 3.

jrtc27 replied:
Huh. That's what it says. But *why*? What possible benefit is there for that?

jrtc27 replied:
I get that in the presence of >32-bit instructions this matters, but in a core that doesn't have such instructions, why would you ever fetch instruction bytes you know you're going to throw away?

ved-rivos replied:
The code put in place here was to be extensible to the forthcoming 48b and 64b instructions.

ved-rivos replied:
But I can update it to fetch 2B instead and it will be equivalent. When 48b and 64b instructions are added, this can be extended.

jrtc27 replied:
Yeah, there's more work to be done to generalise this to >32-bit instructions in the presence of CHERI, so deferring that (with a comment explaining why it's written this way) would be helpful.

jrtc27 replied:
In fact you're going to have issues generalising this to >32-bit instructions even without CHERI. Currently you can ignore the problem that a PMP imposes because the PMP's granularity is 4 byte size and alignment, and therefore given ILEN is 32-bit i.e. 4 bytes you cannot currently ever have a case where fetching these extra bytes (whether because you were already aligned and fetched a whole 4 bytes that contained a 2-byte instruction at the start, or because you were unaligned and fetched 2 + 4 bytes and threw away the final 2 bytes for a resulting 4-byte instruction) would introduce new PMP faulting behaviour. But once you have 48-bit instructions, or more, you can't just fetch things in bigger power-of-two blocks as a generalisation of this, because that might start to straddle PMP boundaries and introduce faults on those wider accesses that would not be there in the 2-byte-at-a-time fetch approach, and you don't want to be able to take instruction fetch faults for instruction bytes you aren't executing.

CHERI is just a case of the same problem due to supporting more fine-grained protection than 4 byte size and alignment, and therefore shows up even when ILEN is 32-bit. And at least for CHERI it's protecting on the virtual address using a capability you have to hand, so you can tell in the frontend when to cut off the fetching (or poison some of the bytes), but for the PMP you can't know until you've translated the address.

I have to say I'm a little concerned that there's no mention of the interaction between Ziccif and the PMP for ILEN > 32 in the Ziccif spec linked to in this PR. Has this been considered at all?

ved-rivos replied:
Thanks.

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -29,33 +40,44 @@ function fetch() -> FetchResult = {
-                  Err(e)  => F_Error(e, PC_hi),
-                  Ok(ihi) => F_Base(append(ihi, ilo))
-                }
+  // An implementation with the Ziccif extension fetches instructions in a manner
```
fetching byte by byte is also a valid implementation. While there are many valid implementations, there is now a specified implementation. Among the many valid but adhoc implementations we can retain the one specified implementation.

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -29,33 +40,44 @@ function fetch() -> FetchResult = {
-          else {
-            /* fetch PC check for the next instruction granule */
-            let PC_hi = PC + 2;
-            match ext_fetch_check_pc(PC, PC_hi) {
```
The hook still exists. Please see read_insn_bytes function.

ved-rivos replied:
I see I have an error. The read_insn_bytes should invoke the hook with params (vaddr, vaddr+num_bytes).

ved-rivos commented:
> 
> Don't worry too much about Lean backend failures, those of use who are working on Sail to Lean can fix issues, and we can merge PRs that fail this particular job.

Thanks @Alasdair. I wanted to make sure that it wasnt my code that caused a breakage.

ved-rivos commented:
> @ved-rivos I believe I fixed the issue in the Lean backend you were running into in #1272. This should pass CI if you rebase.

@jordancarlin Thanks. I rebased the PR.

ved-rivos commented:
@jordancarlin unfortunately CI still fails

```
✖ [62/101] Building LeanRV64D.RiscvSysRegs (5.7s)
trace: .> LEAN_PATH=/home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/lib/lean /home/runner/.elan/toolchains/leanprover--lean4-nightly---nightly-2025-08-02/bin/lean --tstack=400000 /home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/LeanRV64D/RiscvSysRegs.lean -o /home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/lib/lean/LeanRV64D/RiscvSysRegs.olean -i /home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/lib/lean/LeanRV64D/RiscvSysRegs.ilean -c /home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/ir/LeanRV64D/RiscvSysRegs.c --setup /home/runner/work/sail-riscv/sail-riscv/build/model/Lean_RV64D/.lake/build/ir/LeanRV64D/RiscvSysRegs.setup.json --json
error: LeanRV64D/RiscvSysRegs.lean:696:58: failed to prove termination, possible solutions:
  - Use `have`-expressions to prove the remaining goals
  - Use `termination_by` to specify a different well-founded relation
  - Use `decreasing_by` to specify your own tactic for discharging this kind of goal
⊢ max (currentlyEnabled_measure Ext_Ziccif) 0 < currentlyEnabled_measure Ext_Ziccid
error: Lean exited with code 1
error: build failed
Some required targets logged failures:
- LeanRV64D.RiscvSysRegs
```

jordancarlin commented:
Lean is unable to track dependencies between extensions, so they need to be explicitly spelled out in the `currentlyEnabled_measure` function in `riscv_termination.sail`. I just pushed an update to this branch to add that for Ziccid.

ved-rivos commented:
> Lean is unable to track dependencies between extensions, so they need to be explicitly spelled out in the `currentlyEnabled_measure` function in `riscv_termination.sail`. I just pushed an update to this branch to add that for Ziccid.

Thank you. BTW, what is Lean?

jordancarlin commented:
Lean is one of the more actively used theorem prover backends that Sail is able to generate code for. See https://lean-lang.org. @ineol or @tobiasgrosser could probably tell you more.

jordancarlin reviewed (changes requested):
A couple of minor nits, but otherwise LGTM.

ved-rivos commented on README.md:
```diff
@@ -126,10 +126,11 @@ For booting operating system images, see the information under the
 - Svinval extension for fine-grained address-translation cache invalidation, v1.0
 - Svrsw60t59b extension for PTE reserved-for-software bits 60-59, v1.0
 - Physical Memory Protection (PMP)
+- Ziccif Extension for Instruction-Fetch Atomicity, v1.0
```
done

ved-rivos commented on model/riscv_extensions.sail:
```diff
@@ -116,6 +116,16 @@ function clause hartSupports(Ext_Zihpm) = config extensions.Zihpm.supported
+mapping clause extensionName = Ext_Ziccif <-> "ziccif"
+function clause hartSupports(Ext_Ziccif) = config extensions.Ziccif.supported
+function clause currentlyEnabled(Ext_Ziccif) = hartSupports(Ext_Ziccif)
+function clause currentlyEnabled(Ext_Ziccid) = hartSupports(Ext_Ziccid) & currentlyEnabled(Ext_Ziccif)
```
Added a function to check for instruction dependencies. I have included a couple as example but that function needs to be filled out fully.

ved-rivos commented on model/riscv_fetch.sail:
```diff
@@ -11,11 +11,23 @@
+      match translateAddr(Virtaddr(vaddr), InstructionFetch()) {
+        Err(e, _)     => Err(F_Error(e, vaddr)),
+        Ok(paddr,  _) =>
+          match mem_read(InstructionFetch(), paddr, num_bytes, false, false, false) {
```
@pmundkur mem_read invokes mem_read_priv. mem_read_priv invokes mem_read_priv_meta. mem_read_priv_meta invokes checked_mem_read. checked_mem_read does phys_access_check, which invokes pmpCheck, and if successful invokes phys_mem_read. So I think mem_read() is the right function to invoke here.
